### PR TITLE
Wok 450 Color Extraction simple version

### PR DESCRIPTION
**Pull request by VictorHarbo**

### CODE REVIEW

File: src/main/openapi/ds-image-analysis-openapi_v1.yaml [hunk 1]
```diff
@@ -46,13 +41,68 @@ paths:
         '200':
           description: 'The dominant colors'
```
Comment (tokee) [inline src/main/openapi/ds-image-analysis-openapi_v1.yaml]: Just "color". Same with line 23.

File: src/main/openapi/ds-image-analysis-openapi_v1.yaml [hunk 2]
```diff
@@ -46,13 +41,68 @@ paths:
...
+            text/plain:
+              schema:
+                type: string
+                example: "#000000"
+
+  /colorcount:
+    post:
+      tags: 
+        - 'Image Analysis'
+      summary: 'Count the colors in the given image.'
```
Comment (tokee) [inline src/main/openapi/ds-image-analysis-openapi_v1.yaml]: Please add an 'unique' here (and in line 70). We need to be precise in the user-facing parts of our applications.

File: src/main/java/dk/kb/image/api/v1/impl/ImageAnalysisApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        } catch (Exception e){
+            throw handleException(e);
+        }
+        try { 
+            // Show download link in Swagger UI, inline when opened directly in browser
+            setFilename("output", false, false);
+            return output -> output.write(Facade.getGreyscale(img)); 
+        } catch (Exception e){
+            throw handleException(e);
+        }
```
Comment (tokee) [inline src/main/java/dk/kb/image/api/v1/impl/ImageAnalysisApiServiceImpl.java]: No need for separate try-catch: Just wrap it in a single one.

Comment (tokee) [reply src/main/java/dk/kb/image/api/v1/impl/ImageAnalysisApiServiceImpl.java]: Same goes for the other methods.

File: src/main/java/dk/kb/image/Facade.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+import javax.imageio.ImageIO;
+import javax.swing.plaf.ColorUIResource;
+
+import com.ctc.wstx.cfg.OutputConfigFlags;
+
+import java.awt.image.BufferedImage;
+import java.awt.Color;
+
+public class Facade {
+    public static byte[] getGreyscale(BufferedImage img) throws IOException{
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java:28]: Please add JavaDoc for all `public` methods (also fine for `private`, but less important). For the greyscale this should include a general description (as always) and a specific description of what is returned (a JPEG file as bytes) as that is not clear from the method signature.

File: src/main/java/dk/kb/image/Facade.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        int height = img.getHeight();
+
+        // Nested for loop that adds unique colors to the hashset colors
+        for(int y = 0; y < height; y++) {
+            for(int x = 0; x < width; x++) {
+                int pixel = img.getRGB(x, y);     
+                colors.add(pixel);
+            }
+        }
+        String result = "There are " + Integer.toString(colors.size()) + " colors in this picture.";
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: "unique colors"

File: src/main/java/dk/kb/image/Facade.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        return result;
+    }
+
+    public static String colorDistance(BufferedImage img){
+        // Define simple buckets
+        int[] buckets = defineSimpleBuckets();
+        // Count pixels and add to closest bucket
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: "add 1"

File: src/main/java/dk/kb/image/Facade.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        int[] bucketCount = countBucketsForImg(img, buckets);
+        // Get best bucket
+        int bestBucket = getBestBucket(bucketCount);
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: Maybe "largest" or something like that would be better than "best"? "best is very subjective.

File: src/test/java/dk/kb/image/FacadeTest.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FacadeTest {
+    private Logger log = LoggerFactory.getLogger(this.toString());
+
+
+    // method to validate that image is greyscale
+    public static boolean isGreyscale(BufferedImage image){
```
Comment (tokee) [inline src/test/java/dk/kb/image/FacadeTest.java:55]: This method is a helper, along with `canCountColors` so they should be _after_ the primary methods (the ones annotated with `@Test`).

File: src/test/java/dk/kb/image/FacadeTest.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                blue = (pixel) & 0xff;
+                // Check that colors are equal to eachother
+                // If not equal, the image contains color
+                if (red != green || green != blue ) return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean canCountColors(BufferedImage image){
```
Comment (tokee) [inline src/test/java/dk/kb/image/FacadeTest.java]: This looks like a near `1:1` re-implementation of the Facade method. So it does not really test anything. As the input image does not change, maybe use another tool (ImageMagick/Gimp) to count the colors and simply put that number into the `testColorCount`-method directly?

File: src/test/java/dk/kb/image/FacadeTest.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    public void testColorCount() throws IOException {
+        BufferedImage img;
+        img = ImageIO.read(Resolver.resolveStream("flower.jpg"));
+        log.info(Facade.getColorCount(img));
+        assertTrue(canCountColors(img));
+        log.info("Colors get counted"); 
+    }
+
+    @Test
+    public void testColorDistance() throws IOException{
```
Comment (tokee) [inline src/test/java/dk/kb/image/FacadeTest.java]: No asserts here, so it only tests if the `colorDistance` method succeeds, not if the result is usable.

File: src/main/java/dk/kb/image/Facade.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            for(int x = 0; x < width; x++) {
+                int pixel = img.getRGB(x, y);     
+                colors.add(pixel);
+            }
+        }
+        String result = "There are " + Integer.toString(colors.size()) + " colors in this picture.";
+        return result;
+    }
+
+    public static String colorDistance(BufferedImage img){
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: Maybe not a very telling method name?

**Review by tokee — CHANGES_REQUESTED** [review]
Overall very nice! Nothing to say abot the raw functionality,.

File: src/test/java/dk/kb/image/FacadeTest.java (hunk 9 above)
Comment (VictorHarbo) [reply src/test/java/dk/kb/image/FacadeTest.java]: This is indeed very close to a re-implementation. I believe I wrote this test without fully understanding how to use unit testing. I also believe that I made this test to do some kind of sanity check.
In retrospect this test might not be needed at all?

File: src/main/java/dk/kb/image/Facade.java (hunk 4 above)
Comment (VictorHarbo) [reply src/main/java/dk/kb/image/Facade.java:28]: JavaDoc added for all methods

File: src/test/java/dk/kb/image/FacadeTest.java (hunk 8 above)
Comment (VictorHarbo) [reply src/test/java/dk/kb/image/FacadeTest.java:55]: Has been moved

File: src/main/java/dk/kb/image/Facade.java [hunk 12]
```diff
@@ -55,6 +61,12 @@ public static byte[] getGreyscale(BufferedImage img) throws IOException{
         return baos.toByteArray();
     }
 
+    /**
+     * Get the amount of unique colors from input image.
+     * @param img
+     * @return: Returns a string with the count of unique colors form the input image.
+     * @throws IOException
+     */
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: I should have given you better pointers/explanation of Java Doc, sorry. Taking this method as an example:
* `Get the amount of unique colors from input image.`: This provides the extra information "unique", which is a good: It adds to the information from the method name.
* `@param img`: When there is nothing to write, don't state the param at all.
* `@return: Returns a string with the count of unique colors form the input image.`: The "Returns" is redundant and the recommended style (https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html) is that "return" is implied, so `@return a string with the count of unique colors from the input image` would fit the style better.
* `@throws IOException`: Same comment as for `img`: Only state it if an explanation follows.

**Review by tokee — CHANGES_REQUESTED** [review]
More JavaDoc talk.

File: src/test/java/dk/kb/image/FacadeTest.java (hunk 9 above)
Comment (tokee) [reply src/test/java/dk/kb/image/FacadeTest.java]: It is needed if we care about the result being correct :-)

File: src/main/java/dk/kb/image/Facade.java [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+        return baos.toByteArray();
+    }
+
+    /**
+     * Get the amount of unique colors from input image.
+     * @param img
+     * @return: Returns a string with the count of unique colors form the input image.
+     * @throws IOException
+     */
+    public static String getColorCount(BufferedImage img) throws IOException{
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: This should return an `int` so that is can be used programmatically.

**Review by tokee — CHANGES_REQUESTED** [review]

File: src/main/java/dk/kb/image/Facade.java [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+            Color.BLUE.getRGB(),
+            Color.YELLOW.getRGB(),
+            Color.CYAN.getRGB(),
+            Color.MAGENTA.getRGB()
+            };
+    }
+
+    /**
+     * Loop through pixels of input image, get RGB color for pixel and +1 to bucket closest to pixel color.
+     * @param buckets: Integer array of bucket colors.
```
Comment (tokee) [inline src/main/java/dk/kb/image/Facade.java]: Minor correction: No color after the param name (same as for `returns`).

File: src/main/openapi/ds-image-analysis-openapi_v1.yaml [hunk 15]
```diff
@@ -44,15 +39,70 @@ paths:
...
+                    description: 'The image to analyse'
+                    type: string
+                    format: binary
+      responses:
+          '200':
+            description: 'Number of unique colors in image:'
+            content:
+              text/plain:
+                schema:
+                  type: integer
```
Comment (tokee) [inline src/main/openapi/ds-image-analysis-openapi_v1.yaml:74]: Remember to stats the `format` when using an integer (`int32` is fine for unique colors).

**Review by tokee — CHANGES_REQUESTED** [review]

File: src/main/openapi/ds-image-analysis-openapi_v1.yaml (hunk 15 above)
Comment (VictorHarbo) [reply src/main/openapi/ds-image-analysis-openapi_v1.yaml:74]: Format has been added

**Comment by tokee** [comment]
Hooray!

**Pull request merged**